flang/lib/Lower/PFTBuilder.cpp

```diff
@@ -104,14 +104,53 @@ class PFTBuilder {
                                              stmt.position, stmt.label});
         return false;
       } else if constexpr (std::is_same_v<T, parser::ActionStmt>) {
-        addEvaluation(
-            makeEvaluationAction(stmt.unwrapped, stmt.position, stmt.label));
-        return true;
+        return std::visit(
+            common::visitors{
+                [&](const common::Indirection<parser::IfStmt> &x) {
+                  convertIfStmt(x.value(), stmt.position, stmt.label);
+                  return false;
+                },
+                [&](const auto &x) {
+                  addEvaluation(lower::pft::Evaluation{
+                      removeIndirection(x), parentVariantStack.back(),
+                      stmt.position, stmt.label});
+                  return true;
+                },
+            },
+            stmt.unwrapped.u);
       }
     }
     return true;
   }
 
+  /// Convert an IfStmt into an IfConstruct, retaining the IfStmt as the
+  /// first statement of the construct.
+  void convertIfStmt(const parser::IfStmt &ifStmt, parser::CharBlock position,
+                     std::optional<parser::Label> label) {
+    // Generate a skeleton IfConstruct parse node.  Its components are never
+    // referenced.  The actual components are available via the IfConstruct
+    // evaluation's nested evaluationList, with the ifStmt in the position of
+    // the otherwise normal IfThenStmt.
+    enterConstructOrDirective(parser::IfConstruct{
```

Comment:
Can we leave a comment that this is a less than ideal code pattern that we don't really want to see per se? While it's reasonable to make progress and given that we don't have a true AST, etc., I don't want people in the future to think that substituting mock-up parse trees is best practice. (Some day maybe/hopefully the PFT will be replaced with a first-class AST data structure, support syntactic processing, etc.)

Reply:
I've added additional commentary.  I've also switched to single instances of the the new nodes, rather than regenerating them each time.